File: client-sdk-python/amttp/services/kyc.py
"""KYC (Know Your Customer) Service — wraps ``/kyc`` endpoints."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel

from amttp.services.base import BaseService

# ── types ─────────────────────────────────────────────────────────────────────

KYCStatus = Literal["none", "pending", "verified", "rejected", "expired"]
KYCLevel = Literal["none", "basic", "standard", "enhanced"]


class KYCSubmission(BaseModel):
    address: str
    document_type: Literal["passport", "driving_license", "national_id"]
    document_number: str
    first_name: str
    last_name: str
    date_of_birth: str
    nationality: str
    document_front_hash: Optional[str] = None
    document_back_hash: Optional[str] = None
    selfie_hash: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class KYCVerificationResult(BaseModel):
    address: str
    status: str = "none"
    level: str = "none"
    verified_at: Optional[str] = None
    expires_at: Optional[str] = None
    rejection_reason: Optional[str] = None
    required_documents: Optional[List[str]] = None
    provider: Optional[str] = None


class KYCDocument(BaseModel):
    id: str
    type: str
    status: Literal["pending", "verified", "rejected"]
    uploaded_at: str
    verified_at: Optional[str] = None
    expires_at: Optional[str] = None


class KYCRequirements(BaseModel):
    level: str
    required_documents: List[str] = []
    max_transaction_limit: str = ""
    features: List[str] = []


# ── service ───────────────────────────────────────────────────────────────────


class KYCService(BaseService):
    """KYC verification operations."""

    async def submit(self, submission: KYCSubmission) -> KYCVerificationResult:
        """Submit KYC documents for verification."""
        resp = await self._http.post("/kyc/submit", json=submission.model_dump(exclude_none=True))
        resp.raise_for_status()
        data = KYCVerificationResult.model_validate(resp.json())
        self._events.emit("kyc:submitted", {
            "address": submission.address,
            "document_type": submission.document_type,
        })
        return data

    async def get_status(self, address: str) -> KYCVerificationResult:
        """Get KYC status for an address."""
        resp = await self._http.get(f"/kyc/status/{address}")
        resp.raise_for_status()
        return KYCVerificationResult.model_validate(resp.json())

    async def is_verified(self, address: str) -> bool:
        """Check if address is KYC verified."""
        status = await self.get_status(address)
        return status.status == "verified"

    async def get_level(self, address: str) -> str:
        """Get KYC level for an address."""
        status = await self.get_status(address)
        return status.level

    async def upload_document(
        self,
        address: str,
        *,
        doc_type: str,
        content_hash: str,
        mime_type: str,
        encrypted_content: Optional[str] = None,
    ) -> Dict[str, str]:
        """Upload a document for KYC verification."""
        payload: Dict[str, Any] = {
            "type": doc_type,
            "contentHash": content_hash,
            "mimeType": mime_type,
        }
        if encrypted_content:
            payload["encryptedContent"] = encrypted_content
        resp = await self._http.post(f"/kyc/documents/{address}", json=payload)
        resp.raise_for_status()
        result = resp.json()
        self._events.emit("kyc:documentUploaded", {
            "address": address,
            "document_type": doc_type,
            "document_id": result.get("documentId"),
        })
        return result

    async def get_documents(self, address: str) -> List[KYCDocument]:
        """Get uploaded documents for an address."""
        resp = await self._http.get(f"/kyc/documents/{address}")
        resp.raise_for_status()
        return [KYCDocument.model_validate(d) for d in resp.json().get("documents", [])]

    async def get_requirements(self, level: Optional[str] = None) -> List[KYCRequirements]:
        """Get KYC requirements for a specific level."""
        url = f"/kyc/requirements?level={level}" if level else "/kyc/requirements"
        resp = await self._http.get(url)
        resp.raise_for_status()
        return [KYCRequirements.model_validate(r) for r in resp.json().get("requirements", [])]

    async def request_upgrade(self, address: str, target_level: str) -> Dict[str, Any]:
        """Request KYC level upgrade."""
        resp = await self._http.post("/kyc/upgrade", json={
            "address": address, "targetLevel": target_level,
        })
        resp.raise_for_status()
        result = resp.json()
        self._events.emit("kyc:upgradeRequested", {
            "address": address,
            "target_level": target_level,
            "request_id": result.get("requestId"),
        })
        return result

    async def verify_on_chain(self, address: str, chain_id: int) -> Dict[str, Any]:
        """Verify KYC attestation on-chain."""
        resp = await self._http.get(f"/kyc/verify-onchain/{address}?chainId={chain_id}")
        resp.raise_for_status()
        return resp.json()

    async def renew(self, address: str) -> KYCVerificationResult:
        """Renew expiring KYC."""
        resp = await self._http.post(f"/kyc/renew/{address}")
        resp.raise_for_status()
        self._events.emit("kyc:renewed", {"address": address})
        return KYCVerificationResult.model_validate(resp.json())

    async def check_expiry(self, address: str) -> Dict[str, Any]:
        """Check if KYC is expiring soon."""
        resp = await self._http.get(f"/kyc/expiry/{address}")
        resp.raise_for_status()
        return resp.json()

File: client-sdk/src/services/kyc.ts
/**
 * KYC Service
 * Wraps /kyc endpoints for Know Your Customer verification
 */

import { AxiosInstance } from 'axios';
import { EventEmitter } from '../events';
import { BaseService } from './base';

// Local type definitions
export type KYCStatus = 'none' | 'pending' | 'verified' | 'rejected' | 'expired';
export type KYCLevel = 'none' | 'basic' | 'standard' | 'enhanced';

export interface KYCSubmission {
  address: string;
  documentType: 'passport' | 'driving_license' | 'national_id';
  documentNumber: string;
  firstName: string;
  lastName: string;
  dateOfBirth: string;
  nationality: string;
  documentFrontHash?: string;
  documentBackHash?: string;
  selfieHash?: string;
  metadata?: Record<string, any>;
}

export interface KYCVerificationResult {
  address: string;
  status: KYCStatus;
  level: KYCLevel;
  verifiedAt?: string;
  expiresAt?: string;
  rejectionReason?: string;
  requiredDocuments?: string[];
  provider?: string;
}

export interface KYCDocument {
  id: string;
  type: string;
  status: 'pending' | 'verified' | 'rejected';
  uploadedAt: string;
  verifiedAt?: string;
  expiresAt?: string;
}

export interface KYCRequirements {
  level: KYCLevel;
  requiredDocuments: string[];
  maxTransactionLimit: string;
  features: string[];
}

export class KYCService extends BaseService {
  constructor(http: AxiosInstance, events: EventEmitter) {
    super(http, events);
  }

  /**
   * Submit KYC documents for verification
   */
  async submit(submission: KYCSubmission): Promise<KYCVerificationResult> {
    const response = await this.http.post<KYCVerificationResult>('/kyc/submit', submission);
    
    this.events.emit('kyc:submitted', {
      address: submission.address,
      documentType: submission.documentType
    });

    return response.data;
  }

  /**
   * Get KYC status for an address
   */
  async getStatus(address: string): Promise<KYCVerificationResult> {
    const response = await this.http.get<KYCVerificationResult>(`/kyc/status/${address}`);
    return response.data;
  }

  /**
   * Check if address is KYC verified
   */
  async isVerified(address: string): Promise<boolean> {
    const status = await this.getStatus(address);
    return status.status === 'verified';
  }

  /**
   * Get KYC level for an address
   */
  async getLevel(address: string): Promise<KYCLevel> {
    const status = await this.getStatus(address);
    return status.level;
  }

  /**
   * Upload document for KYC verification
   */
  async uploadDocument(address: string, document: {
    type: 'document_front' | 'document_back' | 'selfie' | 'proof_of_address';
    contentHash: string;
    mimeType: string;
    encryptedContent?: string;
  }): Promise<{ documentId: string; status: string }> {
    const response = await this.http.post<{ documentId: string; status: string }>(
      `/kyc/documents/${address}`,
      document
    );
    
    this.events.emit('kyc:documentUploaded', {
      address,
      documentType: document.type,
      documentId: response.data.documentId
    });

    return response.data;
  }

  /**
   * Get uploaded documents for an address
   */
  async getDocuments(address: string): Promise<KYCDocument[]> {
    const response = await this.http.get<{ documents: KYCDocument[] }>(`/kyc/documents/${address}`);
    return response.data.documents;
  }

  /**
   * Get KYC requirements for a specific level
   */
  async getRequirements(level?: KYCLevel): Promise<KYCRequirements[]> {
    const url = level ? `/kyc/requirements?level=${level}` : '/kyc/requirements';
    const response = await this.http.get<{ requirements: KYCRequirements[] }>(url);
    return response.data.requirements;
  }

  /**
   * Request KYC level upgrade
   */
  async requestUpgrade(address: string, targetLevel: KYCLevel): Promise<{
    requestId: string;
    requiredDocuments: string[];
    status: string;
  }> {
    const response = await this.http.post<{
      requestId: string;
      requiredDocuments: string[];
      status: string;
    }>('/kyc/upgrade', { address, targetLevel });
    
    this.events.emit('kyc:upgradeRequested', {
      address,
      targetLevel,
      requestId: response.data.requestId
    });

    return response.data;
  }

  /**
   * Verify KYC attestation on-chain
   */
  async verifyOnChain(address: string, chainId: number): Promise<{
    verified: boolean;
    attestationHash?: string;
    verifiedAt?: string;
    level?: KYCLevel;
  }> {
    const response = await this.http.get<{
      verified: boolean;
      attestationHash?: string;
      verifiedAt?: string;
      level?: KYCLevel;
    }>(`/kyc/verify-onchain/${address}?chainId=${chainId}`);
    
    return response.data;
  }

  /**
   * Renew expiring KYC
   */
  async renew(address: string): Promise<KYCVerificationResult> {
    const response = await this.http.post<KYCVerificationResult>(`/kyc/renew/${address}`);
    
    this.events.emit('kyc:renewed', { address });
    
    return response.data;
  }

  /**
   * Check if KYC is expiring soon
   */
  async checkExpiry(address: string): Promise<{
    isExpiring: boolean;
    expiresAt?: string;
    daysRemaining?: number;
  }> {
    const response = await this.http.get<{
      isExpiring: boolean;
      expiresAt?: string;
      daysRemaining?: number;
    }>(`/kyc/expiry/${address}`);
    
    return response.data;
  }
}
